web/src/main/java/org/phoenixctms/ctsms/web/component/datatable/ColumnManagementBean.java

```diff
@@ -0,0 +1,233 @@
+package org.phoenixctms.ctsms.web.component.datatable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.faces.application.FacesMessage;
+import javax.faces.component.html.HtmlOutputText;
+import javax.faces.event.ActionEvent;
+
+import org.phoenixctms.ctsms.exception.AuthenticationException;
+import org.phoenixctms.ctsms.exception.AuthorisationException;
+import org.phoenixctms.ctsms.exception.ServiceException;
+import org.phoenixctms.ctsms.vo.DataTableColumnVO;
+import org.phoenixctms.ctsms.web.util.JSValues;
+import org.phoenixctms.ctsms.web.util.MessageCodes;
+import org.phoenixctms.ctsms.web.util.Messages;
+import org.phoenixctms.ctsms.web.util.WebUtil;
+import org.primefaces.component.column.Column;
+import org.primefaces.context.RequestContext;
+
+public class ColumnManagementBean {
+
+	//keep filters when hiding columns:
+	public final static boolean FILTER_INVISIBLE = false; // true;
+	private ConcurrentHashMap<String, Map<String, String>> cache;
+
+	public ColumnManagementBean() {
+		cache = new ConcurrentHashMap<String, Map<String, String>>();
+	}
+
+	public Map<String, String> getVisibleMap(String dataTableClientId) {
+		return getVisibleMap(dataTableClientId, null);
+	}
+
+	private void clearVisibleMap(String dataTableClientId) {
+		getVisibleMap(dataTableClientId, new ArrayList<DataTableColumnVO>());
+	}
+
+	private void reloadVisibleMap(String dataTableClientId) {
+		cache.remove(dataTableClientId);
+		getVisibleMap(dataTableClientId, null);
+	}
+
+	public static void resetVisibleMap(String dataTableId) {
+		DataTable dataTable = (DataTable) WebUtil.findComponentById(dataTableId);
+		if (dataTable != null) {
+			WebUtil.getSessionScopeBean().getColumnManager().clearVisibleMap(":" + dataTable.getClientId());
+		}
+	}
+
+	private void updateVisibleMap(String dataTableClientId, Collection<DataTableColumnVO> tableColumnVOs) {
+		getVisibleMap(dataTableClientId, tableColumnVOs);
+	}
+
+	private Map<String, String> getVisibleMap(String dataTableClientId, Collection<DataTableColumnVO> tableColumnVOs) {
+		if (tableColumnVOs == null && cache.containsKey(dataTableClientId)) {
+			return cache.get(dataTableClientId);
+		} else {
+			if (tableColumnVOs == null) {
+				try {
+					tableColumnVOs = WebUtil.getServiceLocator().getUserService().getDataTableColumns(WebUtil.getAuthentication(), WebUtil.getUser().getId(), dataTableClientId,
+							null);
+				} catch (ServiceException e) {
```

Comment:
see https://github.com/phoenixctms/ctsms/blob/master/core/src/main/java/org/phoenixctms/ctsms/intercept/ErrorLogger.java 

not topcal for this particular change, it can basically only fail if the db is offline or permissions are not set.